compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp

```diff
@@ -0,0 +1,460 @@
+
+  // Check that yielded value is an arith.add op, and is accumulating
+  Operation *addOp = yieldedValue.getDefiningOp();
+  if (!addOp || addOp->getNumOperands() != 2) {
+    LLVM_DEBUG(llvm::dbgs() << "no arith.add body op, wrong numOperands\n");
+    return false;
+  }
+  if (!matchPattern(addOp, m_Op<arith::AddFOp>()) &&
+      !matchPattern(addOp, m_Op<arith::AddIOp>())) {
+    LLVM_DEBUG(llvm::dbgs() << "no arith.add body op\n");
+    return false;
+  }
+  Value add0 = addOp->getOperand(0);
+  Value add1 = addOp->getOperand(1);
+  if (!(add0.isa<BlockArgument>() &&
+        add0.cast<BlockArgument>().getArgNumber() == numInputs) &&
+      !(add1.isa<BlockArgument>() &&
+        add1.cast<BlockArgument>().getArgNumber() == numInputs)) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "arith.add body op not accumulating on output\n");
+    return false;
+  }
+
+  // Check that the producer of the add is an arith.mul op
+  Operation *mulOp;
+  if (add0.isa<BlockArgument>()) {
+    mulOp = add1.getDefiningOp();
+  } else {
+    mulOp = add0.getDefiningOp();
+  }
+  if (!mulOp || mulOp->getNumOperands() != 2) {
+    LLVM_DEBUG(llvm::dbgs() << "no arith.mul body op, wrong numOperands\n");
+    return false;
+  }
+  if (!matchPattern(mulOp, m_Op<arith::MulFOp>()) &&
+      !matchPattern(mulOp, m_Op<arith::MulIOp>())) {
+    LLVM_DEBUG(llvm::dbgs() << "no arith.mul body op\n");
+    return false;
+  }
+  Value mul0 = mulOp->getOperand(0);
+  Value mul1 = mulOp->getOperand(1);
+  if (mul0.isa<BlockArgument>() && mul1.isa<BlockArgument>()) {
+    return true;
+  }
+
+  // Check that non block args come from arith.ext ops
+  if (!mul0.isa<BlockArgument>()) {
+    Operation *extOp0 = mul0.getDefiningOp();
+    if (!extOp0 || extOp0->getNumOperands() != 1) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "no arith.ext body op on in0, wrong numOperands\n");
+      return false;
+    }
+    if (!matchPattern(extOp0, m_Op<arith::ExtUIOp>()) &&
+        !matchPattern(extOp0, m_Op<arith::ExtSIOp>()) &&
+        !matchPattern(extOp0, m_Op<arith::ExtFOp>())) {
+      LLVM_DEBUG(llvm::dbgs() << "no arith.ext body op on in0\n");
+      return false;
+    }
+    Value ext0 = extOp0->getOperand(0);
+    if (!ext0 || !ext0.isa<BlockArgument>()) {
+      return false;
+    }
+  }
+  if (!mul1.isa<BlockArgument>()) {
+    Operation *extOp1 = mul1.getDefiningOp();
+    if (!extOp1 || extOp1->getNumOperands() != 1) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "no arith.ext body op on in1, wrong numOperands\n");
+      return false;
+    }
+    if (!matchPattern(extOp1, m_Op<arith::ExtUIOp>()) &&
+        !matchPattern(extOp1, m_Op<arith::ExtSIOp>()) &&
+        !matchPattern(extOp1, m_Op<arith::ExtFOp>())) {
+      LLVM_DEBUG(llvm::dbgs() << "no arith.ext body op on in1\n");
+      return false;
+    }
+    Value ext1 = extOp1->getOperand(0);
+    if (!ext1 || !ext1.isa<BlockArgument>()) {
+      return false;
+    }
+  }
```

Comment:
let's have a helper function, and use it here. Also, we can use `CastOpInterface` for matching. E.g.,

```cpp
bool isCastOfBlockArgument(Operation *op) {
  return isa<CastOpInterface>(op) && op->getNumOperands() == 1 &&
         isa<BlockArgument>(op->getOperand(0));
}
```